### PR TITLE
fix(behavior_path_planner): fix velocity planning at avoidance

### PR DESCRIPTION
**Pull request by rej55**
Signed-off-by: Fumiya Watanabe <rej55.g@gmail.com>

## Description

In the avoidance module, the planned maximum velocity is decided from the velocity of ego vehicle.
So, the planned maximum velocity changes dynamically, which causes some issues such as the target velocity is getting higher as the vehicle accelerates. The motivation of `modifyPathVelocityToPreventAccelerationOnAvoidance()` is to prevent accelerating while avoidance, thus the target velocity must be set statistically before avoidance motion is performed.

In this PR, I changed to plan the maximum velocity from the velocity of ego vehicle at starting avoidance.

Please check that the maximum velocity in avoidance state does not change dynamically with this PR.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -2176,6 +2176,9 @@ void AvoidanceModule::addShiftPointIfApproved(const AvoidPointArray & shift_poin
     const size_t prev_size = path_shifter_.getShiftPointsSize();
     addNewShiftPoints(path_shifter_, shift_points);
 
+    // update ego velocity
+    ego_velocity_starting_avoidance_ = getEgoSpeed();
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2180]: @rej55  `addShiftPointIfApproved()` may be called long time before the avoidance, using the vehicle speed at this point to limit the acceleration does not make sense... does it?

**Comment by rej55** [comment]
@TakaHoribe @tkimura4 

I fixed the calculation of avoidance velocity like following:
- Update ego velocity if the ego is far from the shift point (more than 3.0 sec)
- Set avoidance velocity which has weak acceleration from the point before the shift point (3.0 sec) to the final point of the path points

How do you think?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -1916,33 +1920,42 @@ void AvoidanceModule::modifyPathVelocityToPreventAccelerationOnAvoidance(Shifted
...
+
+  // calc index and velocity to NO_ACCEL_TIME_THR
+  const auto v0 = *ego_velocity_starting_avoidance_ptr_;
+  auto vmax = 0.0;
+  size_t insert_idx = ego_idx;
+  for (size_t i = ego_idx; i < target_idx; ++i) {
+    const auto s =
+      avoidance_data_.arclength_from_ego.at(target_idx) - avoidance_data_.arclength_from_ego.at(i);
+    const auto t = s / std::max(v0, 1.0);
+    if (t < NO_ACCEL_TIME_THR) {
```
Comment (tkimura4) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1941]: Will this if statement always be executed in any case?

(Is there a case where vmax remains at the initial value of 0?)

**Comment by tkimura4** [comment]
image

https://user-images.githubusercontent.com/59680180/177917420-d0572a9e-8fe7-410b-9522-eeeb5e1f6f53.mp4

I confirmed that the maximum speed set by behavior_path_planner during avoidance did not change depending on the vehicle speed.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
Comment (rej55) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1941]: `s` is decreasing with increasing `i`.
However, there is no guarantee to become `t < NO_ACCEL_TIME_THR`.
It should handle the initial value.
I'll fix it.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 3]
```diff
@@ -2195,6 +2208,8 @@ void AvoidanceModule::addShiftPointIfApproved(const AvoidPointArray & shift_poin
       turn_signal_info.turn_signal.command = TurnIndicatorsCommand::ENABLE_RIGHT;
     }
 
+    ego_velocity_starting_avoidance_ptr_.reset();
```
Comment (tkimura4) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2211]: Is it okay if the behavior_path_planner is already in the flow of avoidance and then  another avoidance flow is approved

Comment (TakaHoribe) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2211]: seems this code should be put in a place where we check if deceleration is needed?

Comment (rej55) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2211]: Maybe, is it enough to remove this line?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
Comment (rej55) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1943]: Fixed.
https://github.com/autowarefoundation/autoware.universe/pull/909/commits/667e1c25c724a7c62f7851a2cf75b0110d57b67b

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 3 above)
Comment (rej55) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: I removed this line when I resolve conflict...
https://github.com/autowarefoundation/autoware.universe/pull/909/commits/d3ef9e7ea7935e9e987c5d4b1fa3abca06dc6f14

**Review by tkimura4 — APPROVED** [review]
LGTM

**Pull request merged**